<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Sample implementation</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>



<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<p><strong>gpuEpiScan</strong> provides efficient methods, which were implemented based on GPU, to scan for pairwise epistasis in both case-control study and quantitative studies. It is suitable for genome-wide interaction studies (GWIS) by splitting the computation into manageable chunks. The epistasis methods used by <strong>gpuEpiScan</strong> are adjusted from two published papers (see References below).</p>

<p>Unfortunately, the current <strong>gpuEpiScan</strong> version only support Linux system. We are trying to figure out how to make it work with Windows system. </p>

<p>## Installation
  To install <strong>gpuEpiScan</strong> sucessfully, <code>CUDA</code> and <code>nvcc</code> must be installed. If your <code>CUDA</code> path is not <code>/usr/local/cuda</code> and your <code>nvcc</code> is not global command, a specific <code>CUDA</code> path needs to be given by the user.</p>

<pre><code class="r"># normal install, the default will take CUDA path with /usr/local/cuda 
# or the detected path with shell command &#39;which nvcc&#39;
devtools::install_git(&#39;https://github.com/beibeiJ/gpuEpiScan.git&#39;)

# if the above installation doesn&#39;t work, `CUDA` path needs to be specified
devtools::install_git(&#39;https://github.com/beibeiJ/gpuEpiScan.git&#39;, 
                      build_opts = &#39;--configure-args=YOUR_CUDA_PATH&#39;)
</code></pre>

<p>An alternative way is to download the source code of <strong>gpuEpiScan</strong> then install via</p>

<pre><code class="r"># x.y.z need to be replaced by the real version, e.g., 0.0.1
R CMD INSTALL gpuEpiScan_x.y.z.tar.gz --configure-args=YOUR_CUDA_PATH
</code></pre>

<h2>Sample implementation</h2>

<pre><code class="r"># load package
library(gpuEpiScan)
</code></pre>

<p>First, we generate a small genotype dataset (<code>geno</code>) with sample size of 100 subjects and 1000 variables (e.g., SNPs) as well as a case-control phenotype (<code>p</code>).  </p>

<pre><code class="r">set.seed(123)
geno &lt;- matrix(sample(0:2, size = 100000, 
                      replace = TRUE, prob = c(0.5, 0.3, 0.2)), ncol = 1000)
dimnames(geno) &lt;- list(row = paste0(&quot;IND&quot;, 1:nrow(geno)), 
                       col = paste0(&quot;rs&quot;, 1:ncol(geno)))
p &lt;- c(rep(0, 60), rep(1, 40))
geno[1:5, 1:8]
#&gt;       col
#&gt; row    rs1 rs2 rs3 rs4 rs5 rs6 rs7 rs8
#&gt;   IND1   0   1   0   1   2   0   0   2
#&gt;   IND2   1   0   2   0   0   0   1   0
#&gt;   IND3   0   0   1   1   2   0   0   0
#&gt;   IND4   2   2   1   1   1   0   0   2
#&gt;   IND5   2   0   0   1   0   0   0   0
</code></pre>

<p>To  use <strong>gpuEpiScan</strong>, simply start with the main function <code>gpuEpiScan</code>. There are three mandatory parameters to be set by the user: genotype data, phenotype data and phenotype category (&ldquo;case-control&rdquo; or &ldquo;quantitative&rdquo;). Since the data simulated above is not normalized yet, we need to set <code>scale = TRUE</code>. By passing an integer number to parameter <code>chunksize</code>, the genotype data will be split into several chunks of that size during the calculation. For the example above (using <code>geno</code>), <code>chunksize = 200</code> means each chunk has 20 variables(variants) and the total number of chunks is 5. Moreover, in most cases, the result of epistasis analysis is huge due to the large number of the variable (variants) combinations. To reduce the size of the result file, setting a threshold of the statistical test (<code>zpthres</code>) to have an output cut-off is a practical option. By seting <code>gpuidx = 0</code>, the GPU calculation will run on the GPU device whose index number is equal to 0.</p>

<pre><code class="r">gpuEpiScan(geno1 = geno, 
           pheno = p, 
           phetype = &quot;case-control&quot;,
           outfile = &quot;episcan_1geno_cc&quot;, 
           suffix = &quot;.txt&quot;, 
           zpthres = 0.9, 
           chunksize = 200, 
           scale = TRUE,
           gpuidx = 0)
#&gt; p-value threshold of Z test for output: 0.9 
#&gt; set chunksize: 200 
#&gt; GPU index: 0 
#&gt; [1] &quot;episcan starts:&quot;
#&gt; [1] &quot;Wed Apr  3 14:43:10 2019&quot;
#&gt; [1] &quot;1 chunk loop: Wed Apr  3 14:43:11 2019&quot;
#&gt; [1] &quot;2 chunk loop: Wed Apr  3 14:43:14 2019&quot;
#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.
#&gt; [1] &quot;3 chunk loop: Wed Apr  3 14:43:15 2019&quot;
#&gt; [1] &quot;4 chunk loop: Wed Apr  3 14:43:15 2019&quot;
#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.
#&gt; [1] &quot;5 chunk loop: Wed Apr  3 14:43:16 2019&quot;
#&gt; [1] &quot;GPUepiblaster calculation is over!&quot;
#&gt; [1] &quot;Wed Apr  3 14:43:16 2019&quot;
</code></pre>

<p>The result of <code>gpuEpiScan</code> is stored in the specified file (&ldquo;episcan_1geno_cc.txt&rdquo;). Let&#39;s take a look:</p>

<pre><code class="r">result &lt;- read.table(&quot;episcan_1geno_cc.txt&quot;,
                     header = TRUE,
                     stringsAsFactors = FALSE)
head(result)
#&gt;   SNP1 SNP2     Zscore         ZP
#&gt; 1  rs1  rs2 -1.1095751 0.26718218
#&gt; 2  rs1  rs3 -0.6424023 0.52061202
#&gt; 3  rs2  rs3  0.7232834 0.46950577
#&gt; 4  rs1  rs4 -1.9321601 0.05333975
#&gt; 5  rs2  rs4  1.3648553 0.17229852
#&gt; 6  rs3  rs4 -0.5495350 0.58263837
</code></pre>

<p>In a genome-wide level epistasis study, it is usual to have millions of variables (variants). Analyzing such big data is super time-consuming. The common appoach is to parallelize the task and run the subtasks with High Performance Computing (HPC) techniques, e.g., on a cluster. By splitting genotype data per chromosome, the huge epistasis analysis task can be divided into relatively small tasks. If only 22 chromosomes exist in the initial task, there are 253 ((1+22)*22/2) subtasks after splitting and considering all the combinations of the chromosomes. <code>gpuEpiScan</code> supports two inputs of genotype data by simply passing information to &ldquo;geno1&rdquo; and &ldquo;geno2&rdquo;.  </p>

<pre><code class="r"># simulate data
geno1 &lt;- matrix(sample(0:2, size = 100000, 
                       replace = TRUE, prob = c(0.5, 0.3, 0.2)), ncol = 1000)
geno2 &lt;- matrix(sample(0:2, size = 200000, 
                       replace = TRUE, prob = c(0.4, 0.3, 0.3)), ncol = 2000)
dimnames(geno1) &lt;- list(row = paste0(&quot;IND&quot;, 1:nrow(geno1)), 
                        col = paste0(&quot;rs&quot;, 1:ncol(geno1)))
dimnames(geno2) &lt;- list(row = paste0(&quot;IND&quot;, 1:nrow(geno2)), 
                        col = paste0(&quot;exm&quot;, 1:ncol(geno2)))
p &lt;- rnorm(100)

# scan epistasis
gpuEpiScan(geno1 = geno1,
           geno2 = geno2,
           pheno = p, 
           phetype = &quot;quantitative&quot;,
           outfile = &quot;episcan_2geno_quant&quot;, 
           suffix = &quot;.txt&quot;, 
           zpthres = 0.6, 
           chunksize = 200, 
           scale = TRUE)
#&gt; p-value threshold of Z test for output: 0.6 
#&gt; set chunksize: 200 
#&gt; GPU index: 0 
#&gt; [1] &quot;episcan starts:&quot;
#&gt; [1] &quot;Wed Apr  3 14:43:17 2019&quot;
#&gt; [1] &quot;1 chunk loop: Wed Apr  3 14:43:17 2019&quot;
#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.
#&gt; [1] &quot;2 chunk loop: Wed Apr  3 14:43:18 2019&quot;
#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.
#&gt; [1] &quot;3 chunk loop: Wed Apr  3 14:43:19 2019&quot;
#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.
#&gt; [1] &quot;4 chunk loop: Wed Apr  3 14:43:21 2019&quot;
#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.
#&gt; [1] &quot;5 chunk loop: Wed Apr  3 14:43:22 2019&quot;
#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.

#&gt; Warning in ZtoP(pairdata$Zscore): There is some Z score value over the
#&gt; system length. After converting, the p-value is recorded as 1e-309.
#&gt; [1] &quot;GPUepiHSIC calculation is over!&quot;
#&gt; [1] &quot;Wed Apr  3 14:43:23 2019&quot;
</code></pre>

<h2>References</h2>

<ul>
<li><p>Kam-Thong, T., D. Czamara, K. Tsuda, K. Borgwardt, C. M. Lewis, A. Erhardt-Lehmann, B. Hemmer, et al. (2011). doi:10.1038/ejhg.2010.196.</p></li>
<li><p>Kam-Thong, T., B. Pütz, N. Karbalai, B. Müller-Myhsok, and K. Borgwardt. (2011). doi:10.1093/bioinformatics/btr218</p></li>
</ul>

</body>

</html>
